lib/render/vp2RenderDelegate/basisCurves.cpp

```diff
@@ -0,0 +1,1700 @@
+                    ++ dstIndex; // don't increment the srcIndex
+                    outputValues[dstIndex] = authoredValues[srcIndex];
+                    ++ dstIndex; ++ srcIndex; 
+                }
+                outputValues[dstIndex] = authoredValues[srcIndex];
+                ++dstIndex; // don't increment the srcIndex
+                outputValues[dstIndex] = authoredValues[srcIndex];
+                ++dstIndex; ++ srcIndex;
+            }
+            else {
+                TF_WARN("Unsupported basis: '%s'", basis.GetText());
+            }
+        }
+        TF_VERIFY(srcIndex == authoredValues.size());
+        TF_VERIFY(dstIndex == numVerts);
+    
+        return outputValues;
+    }
+
+    VtValue _BuildCubicIndexArray(const HdBasisCurvesTopology& topology)
+    {
+        /* 
+        Here's a diagram of what's happening in this code:
+
+        For open (non periodic, wrap = false) curves:
+
+          bezier (vStep = 3)
+          0------1------2------3------4------5------6 (vertex index)
+          [======= seg0 =======]
+                               [======= seg1 =======]
+
+
+          bspline / catmullRom (vStep = 1)
+          0------1------2------3------4------5------6 (vertex index)
+          [======= seg0 =======]
+                 [======= seg1 =======]
+                        [======= seg2 =======]
+                               [======= seg3 =======]
+
+
+        For closed (periodic, wrap = true) curves:
+
+           periodic bezier (vStep = 3)
+           0------1------2------3------4------5------0 (vertex index)
+           [======= seg0 =======]
+                                [======= seg1 =======]
+
+
+           periodic bspline / catmullRom (vStep = 1)
+           0------1------2------3------4------5------0------1------2 (vertex index)
+           [======= seg0 =======]
+                  [======= seg1 =======]
+                         [======= seg2 =======]
+                                [======= seg3 =======]
+                                       [======= seg4 =======]
+                                              [======= seg5 =======]
+        */
+        std::vector<GfVec4i> indices;
+
+        const VtArray<int> vertexCounts = topology.GetCurveVertexCounts();
+        bool wrap = topology.GetCurveWrap() == HdTokens->periodic;
+        int vStep;
+        TfToken basis = topology.GetCurveBasis();
+        if(basis == HdTokens->bezier) {
+            vStep = 3;
+        }
+        else {
+            vStep = 1;
+        }
+
+        int vertexIndex = 0;
+        int curveIndex = 0;
+        TF_FOR_ALL(itCounts, vertexCounts) {
+            int count = *itCounts;
+            // The first segment always eats up 4 verts, not just vstep, so to
+            // compensate, we break at count - 3.
+            int numSegs;
+
+            // If we're closing the curve, make sure that we have enough
+            // segments to wrap all the way back to the beginning.
+            if (wrap) {
```

Comment:
The utility functions are copied from HdSt:

https://github.com/PixarAnimationStudios/USD/blob/dev/pxr/imaging/hdSt/basisCurvesComputations.cpp

I think the code can be kept the same for now and when we need to make necessary change to these functions we can then refine them.

Reply:
Then why aren't you linking to the code from HdSt?

Reply:
Then why aren't you linking to the code from HdSt?

Reply:
We are trying to avoid dependency on HdSt. Maybe we can refactor pixar USD code and extract the utility code from HdSt to Hd so it can be reused by other renderers.

Reply:
I will ask in the USD interest group about whether that is possible before any further change in these utility functions.

Reply:
See https://groups.google.com/forum/#!topic/usd-interest/c3g5rdC5r2U for the discussion.

Reply:
@robthebloke I've logged https://github.com/PixarAnimationStudios/USD/issues/1112 for USD as agreed by George in the post.

At this point, I don't think we need to take further action on this part of code. Do you agree?